Enhanced arpeggio - working snapshot

Rikislav opened the pull request:
I'm trying to enhance the arpeggio functions by changing the int8 key semitones array into a structure which can hold data, at present pan and volume.

The chords are now a hardcoded QString, with the future objective to put data into a text or xml file which will be edited through a dedicated controller.

I added to the chords some examples which can show the possibilities (the Pan ones actually change volume and pan), in future releases things should of course be organized better

Here some simple demonstration: https://www.youtube.com/watch?v=e9c6k7iPljQ

PS. I am not a pro c++ programmer, and I guess would need some correction/advice for my empiric solution as to deal with pointers, vectors etc.. Forgive my naive comments, too

Rikislav commented:
Hmmm, I can't find the information why Travis can't compile here while it does on the fork: https://travis-ci.org/Rikislav/lmms

zonkmachine commented:
I'm manually restarting the build jobs that failed. It looks like Travis is a bit grumpy today.

zonkmachine commented:
OK, Travis passed.

Could you skip the formatting of old code altogether? It's really hard to find your changes in there.
And don't open a new PR, just modify this one.
Does older projects with arpeggios open like before?

I'm away from my build machine right now so I can't test this right now.

Rikislav commented:
Older projects using arpeggios should open as before,  I transformed the old chords and added new combinations at the end.

Sorry for the inconvenience of the previous PRs and for the code formatting. The other PRs should be deleted: I pulled this one anew as the previous were too messy and didnt compile on all machines.

I was (still am) too fresh in all this and didn't understand well enough the protocols of collaborating projects. I'll try to be more careful. Should I reformat the two files and commit them over?

zonkmachine commented:
> Should I reformat the two files and commit them over?

Yes. I'd change back to the older formatting and commit. Then squash down to one commit, if you're hip with that, or you can leave it for now and just push to your master.

You now are working from your master branch which isn't wrong but a bit unpractical. I suggest that you create separate branches from master for each project/bug that you're working on. It's probably a good idea to give branche names and commit messages english names.

> The other PRs should be deleted

I don't think there is any function to delete a PR. Just leave it closed.
It takes a bit of patience to get used to git and GitHub. That's all.

Rikislav commented:
I see I'm messing up... Iwas just trying to accomplish

> I suggest that you create separate branches from master for each project/bug that you're working on. It's probably a good idea to give branch names and commit messages english names.

just created the arpeggio branch, where to experiment, and trying to reverse my master to the original lmms/master in which I'm having trouble as am trying to understand reverting, rebasing, resetting... please be patient, I will master it, eventually

I am also trying to

> squash down to one commit

but I fear to flood this common space with unnecessary actions etc...

Rikislav commented:
Ok, formatted the code as best as I could, the compile should be working. I'm switching to another branch to start the next steps:
- externalizing chords to file
- building up an editor

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -37,17 +39,200 @@ class NotePlayHandle;
+	*
+	* @brief The ChordSemiTone struct
+	*/
+	struct ChordSemiTone {
```
The brace should be in the next line according to our coding guidelines

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -37,17 +39,200 @@ class NotePlayHandle;
+	*
+	* Now it's a QVector, with an overloaded [] operator to maintain compatibility
+	*/
+	struct ChordSemiTones : public QVector<ChordSemiTone> {
```
Brace to the next line, too

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -37,17 +39,200 @@ class NotePlayHandle;
+		* @param index
+		* @return
+		*/
+		int8_t inline operator[](int index) const {
```
Brace again

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -90,37 +283,57 @@ class InstrumentFunctionNoteStacking : public Model, public JournallingObject
 
-		int8_t operator [] ( int n ) const
-		{
+		int8_t operator[](int n) const {
```
Brace again

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -90,37 +283,57 @@ class InstrumentFunctionNoteStacking : public Model, public JournallingObject
+		* @param i
+		* @return
+		*/
+		const ChordSemiTone at(int i) const {
```
Brace again

BaraMGB commented on include/InstrumentFunctions.h:
```diff
@@ -90,37 +283,57 @@ class InstrumentFunctionNoteStacking : public Model, public JournallingObject
+		}
+	};
+
+	struct ChordTable : public QVector<Chord> {
```
And Brace

Rikislav commented:
...I must have again made some confusion with this git thing, I'm sure I had formatted all these pointed out. 
As it can be seen in the number of unnecessary commits I struggled a bit to understand how to undo, manage branches, rebase etc.  Now it seems I made another mistake. 
Well, I'll see into it and correct things as soon as I can, what a wonderful lesson!

We have a proverb here, "Never stir shit as it stinks the room", I guess the best thing would have been to start the project anew

BaraMGB commented:
@Rikislav There is no need for. You can push commits until it works. There is no limit for that. Take your time to learn this things. If it's ready we can squash all your commits to one in github before we merge that. It's a good occasion to learn all that git stuff.

BaraMGB commented:
Theres one last brace you forgot (?) ;)

Rikislav commented:
last brace: ..I must be blind, I can't find it!

Meanwhile I'm trying to put data on xml, are there any specific practices I should follow? I thought of putting an "arpeggio.xml" file somewhere (at present I put it at base of the data directory).
Just started, and I see I'll have to change the static approach to chords to make it possible to load / edit the table on the fly.. I'm struggling a bit, with ConfigManager etc. but it's exactly this which makes things interesting and fun!

Considerations: I am just a learner who is having fun and didn't imagine to propose code changes myself. My proposal was the idea, I think its not just the braces that won't be adequate, I guess my coding solutions will be naive, too

BaraMGB commented:
321 in InstrumentFunction.h

Rikislav commented:
Thanks for the tip, can't really understand how that could slip, I looked at the code many times!

BaraMGB commented:
What did you do in the CmakeList.txt? Seems there are only white space changes. This should be reverted, too. 

I'm not really in that arpeggio thing. Can anyone else look at this? @zonkmachine?

zonkmachine commented:
> Can anyone else look at this? @zonkmachine?

I'll look into this in a day or two.

Rikislav commented:
Thanks for the consideration (in this last period couldn't deal with anything unfortunately...)!
CMakeList: should be the same as in the lmms master branch. As I reverted it some space might have slipped, understood the git difftool option only recently and have been using it on another branch as I thought the pull request was ok...

Anyway
I think I will close this pull request to change the project into building up a new native plugin. Here are my considerations:

The goal was to give the possibility to natively create one's own arpeggios, influencing also the key volume, pan, eventually length. 
I was thinking to achieve this by changing the InstrumentFunction core methods without touching the basic points of the application architecture.
Some issues came up which are just too big, for example the impossibility to load on the fly xml data to a static vector object (correct? Couldn't find a workaround yet)

Up to now the branch controller-arpeggio of the Rikislav/lmms fork suits my needs: the data is held in an xml file which is loaded during initialization through an absolute path only available on linux. I am abandoning this branch as I will try to build a new enhanced arpeggio plugin heavily based on the arpeggio and notestacking core code

I think I've finally mastered the main git features and my understanding of lmms architecture, qt and c++ is slowly but inexorably growing... I'll be proposing the change as soon as I have something ready!

Rikislav commented:
UPDATE

As my understanding of the lmms principles grows I'm exploring possibilities for this arpeggio enhancement. Everything is still chaotic and just works to test features and problems. As soon as I settle down for some solution I'll clean up code and propose it.

Meanwhile
- Automation has been added to the single semitone: automation for key, volume and panning
- A new tool plugin has been added to modify/clone arpeggio sequences. (It uses the available material, just a draft up to now)

![rikis arpeggio](https://cloud.githubusercontent.com/assets/13100084/21499370/7a567548-cc35-11e6-83bc-948b91718ae9.png)

Some things seem to work out pretty well, some others don't, with major issues being the performance with complex patterns, the question if arpeggios should be per instrument or project... I'm still exploring if the whole idea is feasible at all.

At present I'm about to try shifting all of the inbound semitone/chord structures to model based classes, let's see how does it turn out

Up to now on my fork Rikislav/lmms the branches are:

- master: the new chord table structure is hardcoded
- enhanced-arpeggio: some code formatting, discarded
- plugin-arpeggio: the chordtable data is stored on an arbitrary xml file. Exploring the possibility of creating an effect plugin, discarded.

- new-arpeggio: the actual work in progress, changing arbitrary xml towards Model architecture, automating single chord semitone data, creating a new tool plugin resembling the FxMixer to edit data.
Things start moderately working

While developing I'll commit the changes to the new-arpeggio branch. As I'm just trying, testing etc. the code is messy. 
If anyone is interested or has suggestions/criticism etc. please let me know and, meanwhile, enjoy the holidays!
Rikislav

Umcaruje commented:
That looks super interesting to me. Keep up the good work!

Rikislav commented:
Thanks!
I'll sure post when things can be revised properly: up to now the enhancements are just fragments set up for testing, unfinished and absolutely messy...

zonkmachine commented:
Cool stuff! I won't have time to try anything out until after 1.2 is released.

Rikislav commented:
So, in this last build there lies my proposal for an enhancement of the stacking/arpeggio functions.

I decided to pull this work in progress snapshot as to gather some advice/criticism/suggestion, especially for the solutions I used as I am not confident enough I'm doing the things right.

The idea was simple: make the chord data editable and automatable. The changes in the Arpeggio and NoteStacking processnote functions were straightforward, the adaptation to the new data structure and the plugin were for my level of skill quite challenging: I am not sure I used the correct solutions, tried to make everything as similar to what is already present.

I opted to maintain a global ChordTable not to unnecessarily burden the data structure. When editing/applying automation to different instruments the user can easily clone/add specific chords to be used

_things done:_

moved Chordtable, Chord and the relative data from InstrumentFunctions.h to new core files and created the new data files
- include/ChordTable.h
- src/core/ChordTable.cpp
- data/presets/ChordTable/ChordTable.ctd
- data/presets/ChordTable/arpeggio.xpf  <-this is not necessary anymore, but holds original data and can be referenced by some chord table functions to initialize data (to be discarded in future)

created the chordtable tool plugin in the dedicated plugin section
- plugins/chordtable_editor/CMakeLists.txt
- plugins/chordtable_editor/chordtable_editor.cpp
- plugins/chordtable_editor/chordtable_editor.h
- plugins/chordtable_editor/logo.png

modified the following files to reflect changes:
- data/presets/CMakeLists.txt
- include/AutomatableSlider.h
- include/DataFile.h
- include/Engine.h
- include/GuiApplication.h
- include/InstrumentFunctionViews.h
- include/InstrumentFunctions.h
- include/PianoRoll.h
- plugins/CMakeLists.txt
- src/core/CMakeLists.txt
- src/core/DataFile.cpp
- src/core/Engine.cpp
- src/core/InstrumentFunctions.cpp
- src/core/Song.cpp
- src/gui/GuiApplication.cpp
- src/gui/MainWindow.cpp
- src/gui/editors/PianoRoll.cpp
- src/gui/widgets/InstrumentFunctionViews.cpp

Up to now I think the code should be looked up by an expert and the appearance of the widget has of course to be improved.
TODO: The "Active" semitone function has to be implemented correctly (I disabled it for now)

Thanks for all, especially the massive work this project is undergoing!

BaraMGB commented:
That's a real big amount of new code. This will be hard to review. When I have time, I look into it. I'm very interested in this. Perhaps you can make a little video of the possibilities with this?

Rikislav commented:
I'll put on something as soon as I can!
R.

2017-01-16 13:09 GMT+01:00 Steffen Baranowsky <notifications@github.com>:

> That's a real big amount of new code. This will be hard to review. When I
> have time, I look into it. I'm very interested in this. Perhaps you can
> make a little video of the possibilities with this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LMMS/lmms/pull/3117#issuecomment-272846304>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMfkNGvJIX_XNK4DSLNgG7nfjCVTPYmwks5rS13xgaJpZM4KwfRf>
> .
>

Rikislav commented:
So, finally I finally had some time to make some videos to properly introduce this complex pull request.

Forgive if they are a bit awkward, it's the first time I'm making this kind of videos.
I put them on YouTube, but the audio somehow scratches, have no more time to fix things, maybe later on...

I saw the incredible amount of work in fixing, changes, proposals you guys are lately dealing with, I'm overall impressed and grateful, so don't feel I am expecting anything, I am having fun and already appreciate all!

[introduction](https://youtu.be/e9c6k7iPljQ)
[some changes to the code](https://youtu.be/gIZYoQZUa3U)
[creating a simple arpeggio combination](https://youtu.be/eAunEcK1dH4)

Rikislav commented:
So, finally I finally had some time to make some videos to properly introduce this complex pull request.

Forgive if they are a bit awkward, it's the first time I'm making this kind of videos.

I saw the incredible amount of work in fixing, changes, proposals you guys are lately dealing with, I'm overall impressed and grateful, so don't feel I am expecting anything, I am having fun and already appreciate all!

[introduction](https://youtu.be/e9c6k7iPljQ)
[some changes to the code](https://youtu.be/gIZYoQZUa3U)
[creating a simple arpeggio combination](https://youtu.be/eAunEcK1dH4)

Rikislav commented:
Back to lmms after a long time!
Adapted to the changes from upstream/master.

Soon I should have some time to ditch the previous horrible videos and make some decent ones as I need to start an audio project where I think I'll be using also this enhancement.
Bye!

zonkmachine commented:
> Back to lmms after a long time!
> Adapted to the changes from upstream/master.

Sorry, we're still busy working on lmms-1.2 so this has been neglected. I had a go and watched the last video above and it looks really promising.

zonkmachine commented:
@Rikislav Unfortunately I can't get this to build now and it looks like it has some merge issues. It's one of my favourite PR's of all time but I have to close it nonetheless. I have updated the original post with a link to the YouTube demonstration and posts it here also for good measure.
https://www.youtube.com/watch?v=e9c6k7iPljQ

Please bump us on our Discord if you need help or feedback on this.

zonkmachine commented:
> I can't get this to build now

I was on the wrong branch somehow, my bad.

Branches master and riki-one builds but you have to tweak some files first. Just the editor won't open on any of those branches. The new arpeggios works though, including panning. Still a close though.
